src/bosh_azure_cpi/lib/cloud/azure/models/disk_id.rb

```diff
@@ -26,103 +25,77 @@ class DiskId
-      array.sort.join(';')
-    end
-
-    def resource_group_name
-      return @default_resource_group_name if @version == VERSION1
-      @id.fetch('resource_group_name', @default_resource_group_name)
+    def self.parse(id_str, default_resource_group_name)
+      id_hash, plain_id = ResObjectId.parse_with_resource_group(id_str, default_resource_group_name)
+      obj_id = new(id_hash, plain_id)
+      obj_id.validate
+      obj_id
     end
 
     def disk_name
-      return @id if @version == VERSION1
-      @id['disk_name']
+      return @plain_id unless @plain_id.nil?
+      @id_hash[DISK_NAME_KEY]
     end
 
     def caching
-      cloud_error('This function should only be called for data disks') unless disk_name.start_with?(DATA_DISK_PREFIX, MANAGED_DATA_DISK_PREFIX)
+      invalid = !disk_name.start_with?(DATA_DISK_PREFIX, MANAGED_DATA_DISK_PREFIX)
+      cloud_error('This function should only be called for data disks') if invalid
 
-      return parse_data_disk_caching_v1(@id) if @version == VERSION1
-      @id['caching']
+      return _parse_data_disk_caching_v1(@plain_id) unless @plain_id.nil?
+      @id_hash[CACHING_KEY]
     end
 
     def storage_account_name
-      cloud_error('This function should only be called for unmanaged disks') if disk_name.start_with?(MANAGED_DATA_DISK_PREFIX)
+      invalid = disk_name.start_with?(MANAGED_DATA_DISK_PREFIX)
+      cloud_error('This function should only be called for unmanaged disks') if invalid
 
-      return parse_storage_account_v1(@id) if @version == VERSION1
-      @id['storage_account_name']
+      return _parse_storage_account_v1(@plain_id) unless @plain_id.nil?
+      @id_hash[STORAGE_ACCOUNT_NAME_KEY]
     end
 
     def validate
-      if @version == VERSION2
-        cloud_error("Invalid disk_name in disk id (version 2) '#{@id}'") if disk_name.nil? || disk_name.empty?
-        cloud_error("Invalid caching in disk id (version 2) '#{@id}'") if caching.nil? || caching.empty?
+      if @plain_id.nil?
+        cloud_error("Invalid disk_name in disk id (version 2) '#{self}'") if disk_name.nil? || disk_name.empty?
+        cloud_error("Invalid caching in disk id (version 2) '#{self}'") if caching.nil? || caching.empty?
         unless resource_group_name.nil?
-          cloud_error("Invalid resource_group_name in disk id (version 2) '#{@id}'") if resource_group_name.empty?
+          cloud_error("Invalid resource_group_name in disk id (version 2) '#{self}'") if resource_group_name.empty?
         end
         if disk_name.start_with?(DATA_DISK_PREFIX)
-          cloud_error("Invalid storage_account_name in disk id (version 2) '#{@id}'") if storage_account_name.nil? || storage_account_name.empty?
+          cloud_error("Invalid storage_account_name in disk id (version 2) '#{self}'") if storage_account_name.nil? || storage_account_name.empty?
         end
       end
     end
 
+    def self._generate_data_disk_name(use_managed_disks)
```

Comment:
Why not use `def _generate_data_disk_name` directly?

Reply:
because it's called in a static method. so this function need to be a static method too.